### ISSUE TITLE
Mounting a volume subpath with many files in it leaks node storage

### ISSUE THREAD

**Issue by Gui13** [kubernetes/kubernetes#94359]
**What happened**:

We have a cronjob copying files from a volume to a backup volume. The backup volume has lots of files in it (92288 at the moment).

The cronjob mounts both volumes with a subpath:

- source volume has ~100 files to be backuped -- storage is azure-files
- backup volume has 92k files (which seems to be the issue there -- the number of files) -- storage is azure-files

Each time we mount the volume, the undelying node loses 130MB of disk storage. Since our cronjob runs every minute, our node fills up at a rate of ~7Gbi per hour.  I'm ok with the cronjob taking up temporary space, but these 130MB seem to be lost.

There seems to be some kind of garbage collection happening at some point which frees a bit of storage on the node, but we had to increase the OS disk storage to accomodate for this leak.

![image](https://user-images.githubusercontent.com/44523/91706408-a4b5b580-eb7e-11ea-8c54-8600598e85a8.png)

**What you expected to happen**:

I expect that mounting volumes doesn't take space in the nodes. Taking up space while running would be ok, but when the pods are GCd the space should be reclaimed.

**How to reproduce it (as minimally and precisely as possible)**:

Create a volume and a cronjob that mounts that volume and runs every minutes. The cronjob can be empty ('sleep 1'). The volume should contain ~100k files.

The cron will run every minute, and leak a some amount of node storage.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.15.11
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): AKS Ubuntu
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by Gui13** [comment]
If this is relevant: our storage class is azure-files but customized so that mount options have `nouser_xattr`. This is because we are using velero as a backup solution (see [here](https://github.com/restic/restic/issues/1800) for why we have to use that).

I'm not sure but this could take up node space if the container has to "store" these attributes on the writable layer of the container.

**Comment by pacoxu** [comment]
I think mounted volume will not use disk of host at all. 

Would you check the image size of your cronjob ? 
Who occupied the 130MB disk?

I tried to mount host path to cronjob with 100k+ files. And no extra disk  is used by jobs.

Another way is that you can run another job to clean old jobs.

**Comment by Gui13** [comment]
Hi @pacoxu : this is what we did to investigate:

First of all: our job is using bash:latest as docker image. We copied a bash script in there consisting of an 'ls' and then a copy of some files. The image in sub-100Mb (I don't have it readily available). The issue should not be the image size because once the image is on the node, it should not be "adding up" when creating a container, am I right?

The context is this:
We run on Azure Kubernetes. Both volumes are ReadWriteMany with a modified `azurefiles` class containing the `nouser_xattr` mount option.

The volumes are the following:

- one volume to be backed up, named `incoming_files` with about 10-50 files in the root of it:

``` bash
$ ls -l | wc -l .
50 # the files to be copied to the backup volume
```

- the backup volume, named `backup` with one subfolder containing ~100k files. This volume is mounted with the `subPath: subfolder` attribute:

```bash
$ ls -l | wc -l .
1 # the 'subfolder' directory
$ ls -l subfolder/ | wc -l 
90000
```

So from our investigations:

- when the job runs _without any volume mount_, NO space is wasted: the disk available doesn't change
- when the job runs _with the volumes declared but NOT mounted_, NO space is wasted
- when the job runs _with just the first volume mounted_ (but this volume is not containing 1000s of files), NO space is wasted
- when the job runs _with the volume with many files mounted BUT without the `subPath:` where the files reside specified_, NO space is wasted
- when the job runs _with the volume with many files mounts WITH the `subPath`specified_, then 130MB of space is wasted each time the job is started.

This is where we are now. The only difference there is between the "leaking" volume and the not leaking volume is that there are many files at its root.

**Comment by Zheaoli** [comment]
I think it's should be an Azure storage CSI driver error? I will try it on my azure cluster later.

/cloud-provider

**Comment by Gui13** [comment]
We DO have CSI installed but we are not using it in the volumes impacted. ( we don't use CSI at all as a matter of fact, this was installed just as a test).

We will uninstall the CSI driver to check whether this solves the issue (in which case the CSI driver will definitely be in cause, but this would mean that the driver impacts non-related volumes?)

**Comment by fredgate** [comment]
We have uninstalled the CSI driver and we still lose 118 MB of disk each time a job is executed (ie each minute).

**Comment by Gui13** [comment]
(@fredgate is a collegue of mine)

**Comment by Gui13** [comment]
Just FYI I have made the very same test on a GCP cluster, and it looks like:

- we lose a bit less of storage at each invocation, but we still lose storage
- the garbage collection mechanism in GCP is a bit more aggressive and cleans up the remnants correctly:

<img width="1623" alt="image" src="https://user-images.githubusercontent.com/44523/92235926-c8e20100-eeb4-11ea-83ba-9d133eeb8dea.png">

The test is setup thusly:

- I have the same volumes as in Azure, but the storage class is an NFS mount instead of azurefiles, since GCP doesn't offer RWMany storage classes
- I have created the same tree structure for the files, and created 100k files in the subpath just like in AKS.

**Comment by Gui13** [comment]
So what I gather from that is that there is a bit of both culprits:

- Azure loses ~1.5Mb per 1000 files in a subpath-mounted volume
- GCP loses ~300Kb per 1000 files in a subpath-mounted volume

Either way, mounting a file-count-heavy subpath yields a net loss of storage on the node.

**Comment by Gui13** [comment]
Hello,

We just lost another node on Azure due to ephemeral storage fill up.
Are we the only people encountering this? 

We mitigated the issue by compressing files daily and putting them in a subdirectory, but this is not sufficient: it seems that Azure never actually cleans up the ephemeral storage.

We don't have the issue on a GCP node where the ephemeral *does* fill up a bit, but is correctly vacuumed at regular intervals. This cleanup doesn't seem to be working on Azure.

**Comment by Gui13** [comment]
Is there any way to help you guys regarding this? 

If the bug bounces between stale & rotten & "fresh" I'm not sure that it will move much.

Are we the only ones impacted by this problem ?

**Comment by Gui13** [comment]
Up ..

**Comment by Gui13** [comment]
Not stale, we still have the issue!
/unlabel stale

**Comment by wzshiming** [comment]
This bug appears to have been addressed now that #101795 has merged.

/close
